crates/ide_completion/src/item.rs

```diff
@@ -29,12 +29,14 @@ pub struct CompletionItem {
     /// Range of identifier that is being completed.
     ///
     /// It should be used primarily for UI, but we also use this to convert
-    /// genetic TextEdit into LSP's completion edit (see conv.rs).
+    /// generic TextEdit into LSP's completion edit (see conv.rs).
     ///
     /// `source_range` must contain the completion offset. `insert_text` should
     /// start with what `source_range` points to, or VSCode will filter out the
     /// completion silently.
     source_range: TextRange,
+    /// Prefix of `source_range` to be considered if insertion is preferred over replacement.
+    insert_range: Option<TextRange>,
```

Comment:
I wonder if we need two fields here? Can we set source_range to the whole identifier and derive the other range using cursor’s position?


Alternatively, should we add something else here? I removed there being some exceptions around insertions handlers in IntelliJ. I think it was this example:


    foo$0bar()


And we are completing foobaz(). The two results are 


    foobaz()
    foobaz()bar()


They differ not only in the range replaced, but in the number of inserted parameters as well.

Reply:
Right, we can calculate the insert_range in the to_proto transforms since we have the cursor position there still, that seems a lot nicer.

Regarding param completion I'm unsure what to do there, I personally have that feature disabled so I don't really know how it even works/what exactly it does 😅

Reply:
So with this it looks like it prefills params just fine when completing a new ident, but when inside of an ident like in your example we just get `foobazbar`, as in no param filling will be given at all 🤔

Reply:
>Edit: Right so it just prefills the params

You might want to try this setting: https://github.com/matklad/config/blob/master/home/.config/Code/User/settings.json#L128

That way, the feature is much less annoying

Reply:
The main reason why I turned those off is because I got annoyed whenever completing function names in stuff like `Option::map($0)` gave me unnecessary parens(maybe that can be made smarter?).

That aside, the behavior right now seems to be that we don't give parentheses completion unless we are typing at the end of an ident. That means forcing a completion in `foobar$0()` with `foobar` will not insert another pair of parens either. I looked through IntelliJ a bit but I don't see their special handling for this, if they do that is.

I'm unsure if this behaviour should be changed in any way?

Reply:
yeah, map issue should be fixed. 

I've looked at intellij, and I indeed don't see special handling. Looks like I am just misremembering.